tests/test_runner.py::test_submit_candidate FAILED

So the current CI is failing due to one workflow's step.
REF: [Python Package -> Build & test - Debian Testing -> Test using ./tests/test_*.py (unittest)](https://github.com/jelmer/janitor/actions/runs/13415979661/job/37476837730)

```console
[...]
+ su postgres -c make test
[...]
PYTHONPATH=/__w/janitor/janitor/py: PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python python3 -m pytest -vv tests
[...]
tests/test_runner.py::test_candidate_invalid_value PASSED                [ 78%]
tests/test_runner.py::test_submit_candidate FAILED                       [ 79%]
tests/test_runner.py::test_submit_unknown_candidate_codebase PASSED      [ 80%]
[...]
____________________________ test_submit_candidate _____________________________

aiohttp_client = <function aiohttp_client.<locals>.go at 0x7fbdd61e99e0>
db = <asyncpg.pool.Pool object at 0x7fbdd608f640>
tmp_path = PosixPath('/tmp/pytest-of-postgres/pytest-0/test_submit_candidate0')

    async def test_submit_candidate(aiohttp_client, db, tmp_path):
        vcs = tmp_path / "vcs"
        vcs.mkdir()
        qp = await create_queue_processor(db, vcs_managers=get_vcs_managers(str(vcs)))
        client = await create_client(aiohttp_client, qp, campaigns=["mycampaign"])
        resp = await client.post(
            "/codebases",
            json=[{"name": "foo", "branch_url": "https://example.com/foo.git"}],
        )
        assert resp.status == 200
        resp = await client.post(
            "/candidates",
            json=[
                {
                    "campaign": "mycampaign",
                    "codebase": "foo",
                    "command": "true",
                }
            ],
        )
        assert resp.status == 200
        [result] = (await resp.json())["success"]
        assert result == {
            "bucket": "default",
            "campaign": "mycampaign",
            "change_set": None,
            "codebase": "foo",
            "estimated_duration": 15.0,
            "offset": 35000.0,
            "queue-id": 1,
            "refresh": False,
        }
    
        resp = await client.post("/active-runs", json={})
>       assert resp.status == 201
E       AssertionError: assert 500 == 201
E        +  where 500 = <ClientResponse(http://127.0.0.1:42877/active-runs) [500 Internal Server Error]>\n<CIMultiDictProxy('Content-Type': 'text/plain; charset=utf-8', 'Content-Length': '55', 'Date': 'Wed, 19 Feb 2025 15:52:00 GMT', 'Server': 'Python/3.13 aiohttp/3.11.12', 'Connection': 'close')>\n.status
[...]
=========================== short test summary info ============================
FAILED tests/test_runner.py::test_submit_candidate - AssertionError: assert 500 == 201
 +  where 500 = <ClientResponse(http://127.0.0.1:42877/active-runs) [500 Internal Server Error]>\n<CIMultiDictProxy('Content-Type': 'text/plain; charset=utf-8', 'Content-Length': '55', 'Date': 'Wed, 19 Feb 2025 15:52:00 GMT', 'Server': 'Python/3.13 aiohttp/3.11.12', 'Connection': 'close')>\n.status
FAILED tests/test_site.py::test_duration - AssertionError: assert '2w1d' == 'f{weeks}w{days}d'
  
  - f{weeks}w{days}d
  + 2w1d
================== 2 failed, 82 passed, 39 warnings in 36.16s ==================
make: *** [Makefile:58: test] Error 1
Error: Process completed with exit code 2.
```

- - -

## Setup

Getting the service setup:

```console
$ podman pull ghcr.io/jelmer/janitor/runner:latest
[...]
$ podman run \
    --name runner \
    --hostname runner \
    --tty \
    --replace \
    --volume /srv/janitor.kali.org/janitor.conf:/janitor/janitor.conf:ro \
    --volume /srv/janitor.kali.org/runner-backup:/janitor/runner-backup:rw \
    --workdir /janitor \
    --network=slirp4netns:allow_host_loopback=true \
    --publish 9911:9911 \
    ghcr.io/jelmer/janitor/runner:latest \
      --config=/janitor/janitor.conf \
      --backup-directory=/janitor/runner-backup \
      --public-vcs-location=https://janitor.example.org/ \
      --public-apt-archive-location=http://127.0.0.1:9935
```

- - - 

## Step 1

**test_runner.py**:

```python
    resp = await client.post(
        "/codebases",
        json=[{"name": "foo", "branch_url": "https://example.com/foo.git"}],
    )
    assert resp.status == 200
```

```console
$ curl -v --data '[{"name":"foo", "branch_url":"https://example.com/foo.git"}]' localhost:9911/codebases; echo
[...]
< HTTP/1.1 200 OK
[...]
{}
$
```

- - -

## Step 2

**test_runner.py**:

```python
    resp = await client.post(
        "/candidates",
        json=[
            {
                "campaign": "mycampaign",
                "codebase": "foo",
                "command": "true",
            }
        ],
    )
    assert resp.status == 200
    [result] = (await resp.json())["success"]
    assert result == {
        "bucket": "default",
        "campaign": "mycampaign",
        "change_set": None,
        "codebase": "foo",
        "estimated_duration": 15.0,
        "offset": 35000.0,
        "queue-id": 1,
        "refresh": False,
    }
```

```console
$ curl -v --data '[{"campaign":"unchanged", "codebase":"foo", "command":"true"}]' localhost:9911/candidates; echo
[...]
< HTTP/1.1 200 OK
[...]
{"success": [{"campaign": "unchanged", "codebase": "foo", "bucket": "default", "change_set": null, "offset": 35000.0, "estimated_duration": 15.0, "queue-id": 8, "refresh": false}], "invalid_command": [], "invalid_value": [], "unknown_campaigns": [], "unknown_codebases": [], "unknown_publish_policies": []}
```

_Using `unchanged` rather than `mycampaign` for the campaign as its defined in `/srv/janitor.kali.org/janitor.conf`._

- - - 

## Step 3

**test_runner.py**:

```python
    resp = await client.post("/active-runs", json={})
    assert resp.status == 201
```

```console
$ curl -v --data '{}' localhost:9911/active-runs; echo
[...]
< HTTP/1.1 500 Internal Server Error
[...]
500 Internal Server Error

* Closing connection 0
Server got itself in trouble
$
```

:(

Now, looking at podman logs:

```plaintext
INFO:aiohttp.access:10.0.2.100 [19/Feb/2025:20:16:09 +0000] "POST /codebases HTTP/1.1" 200 161 "-" "curl/7.88.1"
INFO:aiohttp.access:10.0.2.100 [19/Feb/2025:20:16:16 +0000] "POST /candidates HTTP/1.1" 200 467 "-" "curl/7.88.1"
INFO:root:Opening branch https://example.com/foo.git with ['BzrProber', 'RemoteBzrProber', 'LocalGitProber', 'RemoteGitProber']
ERROR:aiohttp.server:Error handling request from 10.0.2.100
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/dist-packages/janitor/vcs.py", line 181, in open_branch_ext
    return open_branch(vcs_url, possible_transports, probers=probers)
silver_platter.BranchMissing: ('https://example.com/foo.git', 'https://example.com/foo.git/')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_protocol.py", line 480, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_app.py", line 569, in _handle
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp_openmetrics/__init__.py", line 79, in metrics_middleware
    response = await handler(request)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/state.py", line 263, in asyncpg_error_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiozipkin/aiohttp_helpers.py", line 157, in aiozipkin_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/runner.py", line 2347, in handle_assign
    assignment = await next_item(
                 ^^^^^^^^^^^^^^^^
    ...<9 lines>...
    )
    ^
  File "/usr/local/lib/python3.13/dist-packages/janitor/runner.py", line 2582, in next_item
    main_branch = await to_thread_timeout(
                  ^^^^^^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
    )
    ^
  File "/usr/local/lib/python3.13/dist-packages/janitor/runner.py", line 149, in to_thread_timeout
    return await cor
           ^^^^^^^^^
  File "/usr/lib/python3.13/asyncio/tasks.py", line 507, in wait_for
    return await fut
           ^^^^^^^^^
  File "/usr/lib/python3.13/asyncio/threads.py", line 25, in to_thread
    return await loop.run_in_executor(None, func_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/concurrent/futures/thread.py", line 59, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.13/dist-packages/janitor/vcs.py", line 188, in open_branch_ext
    raise _convert_branch_exception(vcs_url, e) from e
          ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/vcs.py", line 142, in _convert_branch_exception
    if e.url not in msg:
       ^^^^^
AttributeError: 'BranchMissing' object has no attribute 'url'
INFO:aiohttp.access:10.0.2.100 [19/Feb/2025:20:19:36 +0000] "POST /active-runs HTTP/1.1" 500 247 "-" "curl/7.88.1"
```

- - - 

Did a bit of fuzzing:

```console
$ curl -v --data '[{}]' localhost:9911/active-runs; echo



ERROR:aiohttp.server:Error handling request from 10.0.2.100
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_protocol.py", line 480, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_app.py", line 569, in _handle
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp_openmetrics/__init__.py", line 79, in metrics_middleware
    response = await handler(request)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/state.py", line 263, in asyncpg_error_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiozipkin/aiohttp_helpers.py", line 157, in aiozipkin_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/runner.py", line 2343, in handle_assign
    assignment_count.labels(worker=json.get("worker")).inc()
                                   ^^^^^^^^
AttributeError: 'list' object has no attribute 'get'
INFO:aiohttp.access:10.0.2.100 [19/Feb/2025:20:16:23 +0000] "POST /active-runs HTTP/1.1" 500 247 "-" "curl/7.88.1"
```

```console
$ curl -v --data '[]' localhost:9911/active-runs; echo



[...]
  File "/usr/local/lib/python3.13/dist-packages/janitor/runner.py", line 2343, in handle_assign
    assignment_count.labels(worker=json.get("worker")).inc()
                                   ^^^^^^^^
AttributeError: 'list' object has no attribute 'get'
INFO:aiohttp.access:10.0.2.100 [19/Feb/2025:20:22:57 +0000] "POST /active-runs HTTP/1.1" 500 247 "-" "curl/7.88.1"
```

```console
$ curl -v --data '' localhost:9911/active-runs; echo



ERROR:aiohttp.server:Error handling request from 10.0.2.100
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_protocol.py", line 480, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_app.py", line 569, in _handle
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp_openmetrics/__init__.py", line 79, in metrics_middleware
    response = await handler(request)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/state.py", line 263, in asyncpg_error_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiozipkin/aiohttp_helpers.py", line 157, in aiozipkin_middleware
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/janitor/runner.py", line 2342, in handle_assign
    json = await request.json()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/dist-packages/aiohttp/web_request.py", line 695, in json
    return loads(body)
  File "/usr/lib/python3.13/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^
  File "/usr/lib/python3.13/json/decoder.py", line 345, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/json/decoder.py", line 363, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
INFO:aiohttp.access:10.0.2.100 [19/Feb/2025:20:22:20 +0000] "POST /active-runs HTTP/1.1" 500 247 "-" "curl/7.88.1"
```

```console
$ curl -v localhost:9911/active-runs; echo

[...]
< HTTP/1.1 200 OK
[...]
[{"queue_id": 8, "id": "66ed6a13-41f2-4d33-ab36-ebb054fdf504", "codebase": "foo", "change_set": null, "campaign": "unchanged", "command": "true", "estimated_duration": 15.0, "current_duration": 0.000214, "start_time": "2025-02-19T20:19:36.482638", "worker": null, "worker_link": null, "vcs": {"branch_url": "https://example.com/foo.git"}, "backchannel": {}, "instigated_context": null, "resume_from": null, "last-keepalive": "2025-02-19T20:19:36", "keepalive_age": 268.774404, "mia": false}, {"queue_id": 1, "id": "1c6c0757-cbc4-4fed-a818-3ba5c47f2de2", "codebase": "foo", "change_set": null, "campaign": "fresh-releases", "command": "true", "estimated_duration": 15.0, "current_duration": 0.000218, "start_time": "2025-02-19T20:18:19.716575", "worker": null, "worker_link": null, "vcs": {"branch_url": "https://example.com/foo.git"}, "backchannel": {}, "instigated_context": null, "resume_from": null, "last-keepalive": "2025-02-19T20:18:19", "keepalive_age": 345.540475, "mia": false}]
```

- - -

I had a look back over the commit history to see when it was last passing: [f30217074aaca5a0013f8ca8360ec9abf6bce802 (2024-08-03)](https://github.com/jelmer/janitor/commit/f30217074aaca5a0013f8ca8360ec9abf6bce802)

--- a/py/janitor/site/__init__.py
+++ b/py/janitor/site/__init__.py
@@ -110,10 +110,10 @@ def update_vars_from_request(vs, request):
 
 def format_duration(duration):
     weeks = duration.days // 7
-    duration.days % 7
-    if weeks:
-        return "f{weeks}w{days}d"
-    if duration.days:
+    days = duration.days % 7
+    if weeks > 0:
+        return f"{weeks}w{days}d"
+    if duration.days > 0:
         return f"{duration.days}d{duration.seconds//(60*60)}h"
     hours = duration.seconds // (60 * 60)
     seconds = duration.seconds % (60 * 60)

